ui/plugin-system/src/components/PluginRegistry/registry-state.ts

```diff
@@ -77,49 +75,53 @@ export function useRegistryState(installedPlugins?: PluginModuleResource[]) {
     return loadedPlugins;
   });
 
-  // Create the register callback to pass to the module's setup function
-  const registerPlugin: RegisterPlugin = useCallback(
-    <Options extends JsonObject>(config: PluginRegistrationConfig<Options>) => {
-      // Just cast to the runtime plugin type that framework code knows about since the `Options` generic argument is
-      // really only known to plugin authors for their type safety when developing plugins in Typescript
-      switch (config.pluginType) {
-        case 'Variable':
-          setPlugins((draft) => {
-            draft.Variable[config.kind] = config.plugin as unknown as VariablePlugin;
-          });
-          return;
-        case 'Panel':
-          setPlugins((draft) => {
-            draft.Panel[config.kind] = config.plugin as unknown as PanelPlugin;
-          });
-          return;
-        case 'GraphQuery':
-          setPlugins((draft) => {
-            draft.GraphQuery[config.kind] = config.plugin as unknown as GraphQueryPlugin;
-          });
-          return;
-        default:
-          const exhaustive: never = config;
-          throw new Error(`Unhandled plugin config: ${exhaustive}`);
-      }
-    },
-    [setPlugins]
-  );
-
-  const registeredModules = useRef(new Set<PluginModule>());
+  const registeredModules = useRef(new Set<unknown>());
   const register = useCallback(
-    (pluginModule: PluginModule): void => {
+    (resource: PluginModuleResource, pluginModule: unknown): void => {
       // De-dupe register calls in case multiple plugin loading boundaries
       // are waiting for the same module in parallel
       if (registeredModules.current.has(pluginModule)) {
         return;
       }
 
-      // Call the setup function and remember it's been registered
-      pluginModule.setup(registerPlugin);
+      // Treat plugin module as JS module with named exports that are each a Plugin
+      const plugins = pluginModule as Record<string, Plugin<JsonObject>>;
+
+      setPlugins((draft) => {
+        // Look for all the plugins specified in the metadata
+        for (const pluginMetadata of resource.spec.plugins) {
+          // Assume that plugins will be exported under the same named export as the kind they handle
```

Comment:
This reminds me, I was thinking earlier this week about if it would make sense to make `kind: 'LineChart'` have an alias of `kind: 'TimeSeries'` somehow. That way we could gradually make that switch instead of introducing the breaking change right away. 

It seems like maybe that wouldn't be very straightforward based on these lines?

Reply:
Good question. I think it would still be pretty easy to do. You'd basically just add to the module exports from `panels-plugin`:

```
export { LineChart as TimeSeries }
```

And then make sure to include an entry in the `plugin.json` for it.